### DISCUSSION TITLE
About Micropython documents

### DISCUSSION THREAD

**Discussion by penitko** [micropython/micropython#13780] [Libraries & Drivers]
I am a Python hobbyist and I have often wondered where to find the documentation for modules, classes, functions, function parameters and function return values.  Eventually I found a way to get Linux desktop python docs to read like:

- python3 -m pydoc -p port_number
- python3 -m pydoc module_name

![desktop-python](https://github.com/micropython/micropython/assets/98764466/4645ab67-5722-43d5-9a05-1a45979b165a)

But is there a similar function in Micropython, e.g. in REPL?
In Thonny the previous one works somehow but I have not managed to get the parameters of the functions visible:

>>> import socket
>>> help(socket)
...
  recvfrom -- <function>
  setsockopt -- <function>
...
![thonny-micropython](https://github.com/micropython/micropython/assets/98764466/ff13bf2b-4381-498a-a2ae-1f1ddd731046)

Thonny only shows the names of the functions in the class, but does not show the parameters and return values of the functions. 
Is there any way to show function parameters and return values?

**Comment by robert-hh** [comment]
There is the MicroPyhthon documentation at https://docs.micropython.org/en/latest/

**Reply by penitko** [reply]
Thanks for the tip but for example the function socket.setsockopt lacks more detailed parameter representation. I wonder where to find them ?
![parameters-missing](https://github.com/micropython/micropython/assets/98764466/736ac898-a2ab-4627-b57d-ef0fd0bcfff9)

**Comment by GitHubsSilverBullet** [comment]
> I am a Python hobbyist and I have often wondered where to find the documentation for modules, classes, functions, function parameters and function return values. Eventually I found a way to get Linux desktop python docs to read like:

Often it's easier to install nginex or apache and put all documentation in a specific »/var/www/html/docs/« subdirectory. Manually build an index.html and voila…
All docs in one place: localhost
Yes, can be easily done with Python, Micropython and lots of other stuff.

**Comment by scruss** [comment]
> But is there a similar function in Micropython, e.g. in REPL?

No. MicroPython is small. There is no room for detailed documentation inside the interpreter.

**Reply by penitko** [reply]
Sure the space of memory of MP devices are restricted in case for documentation purposes. But according to my understanding, somewhere globally should be such a documentation available of MicroPython that covers all that normally has documented with other programming languages, like function's parameter descriptions. But summer is comming and makes hobbyist thinking other things than microcontrollers and their programming.

**Reply by GitHubsSilverBullet** [reply]
The socket module is documented in Micropython. If that's not enough, then one takes a look into the Python3 documentation. There one finds a convenient reference to the underlying OS function/call and all its parameters.
It's all there, but one has to look for it. TANSTAAFL.

**Comment by penitko** [comment]
"The socket module is documented in Micropython. "
Could you give me an example of socket.socket.setsocketopt's function's parameter names and types explainded in micropython's documentation? 
Seems that  you don't understand that I am talking Microrpython's documentation?

**Reply by scruss** [reply]
Not all of it will be in MicroPython's documentation. Some of the way that sockets behave is defined in POSIX documentation. socket options are defined in the Linux/Unix manual page [socket(7)](https://man7.org/linux/man-pages/man7/socket.7.html). There may be more in the source: [micropython/extmod/modnetwork.h at 678707c8b07323c5b914778708a2858387c3b60c · micropython/micropython](https://github.com/micropython/micropython/blob/678707c8b07323c5b914778708a2858387c3b60c/extmod/modnetwork.h#L49)

But, very importantly, the options available to you are *port-specific*. The MicroPython that you are running may not support all/many of the options. 

Also, if you were expecting neat, packaged documentation with complete coverage for every module, you're not going to find it in MicroPython. It's developed by a small team, and there are no people dedicated solely to documentation.

**Reply by penitko** [reply]
Ok, I understand that there are not always enough resources for everything. But now I know that those fully documented documents don't exist.

**Comment by GitHubsSilverBullet** [comment]
> Could you give me an example of socket.socket.setsocketopt's function's parameter names and types explainded in micropython's documentation?
> Seems that you don't understand that I am talking Microrpython's documentation?

It seems that you did not read what I wrote and the idea I tried to convey.
Sockets are sockets are sockets. Not a Micropython invention.
Look into the real Python docs and you'll see.
There's also a **direct LINK** to the manpages for the underlying function call and parameters.

But let me make life easier for you:

https://docs.python.org/3.9/library/socket.html?highlight=setsock#socket.socket.setsockopt

https://manpages.debian.org/setsockopt(2)

**Reply by penitko** [reply]
Ok, now I know for sure that Micropython does not publish all parameters of functions in its documentation. I don't expect to find those docs anywhere anymore. And that's a good thing that in the future I am not stressing myselft with this thing any more.
